qiskit/scheduler/methods/basic.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The most straightforward scheduling methods: scheduling as early or as late as possible."""
+
+from collections import defaultdict
+from typing import List
+
+from qiskit.circuit.measure import Measure
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.extensions.standard.barrier import Barrier
+from qiskit import pulse
+from qiskit.pulse.schedule import Schedule
+
+from qiskit.scheduler.models import ScheduleConfig, CircuitPulseDef
+
+
+def as_soon_as_possible(circuit: QuantumCircuit,
+                        schedule_config: ScheduleConfig) -> Schedule:
+    """
+    Return a pulse Schedule which nails down the timing between circuit element schedules by
+    scheduling pulses as soon as possible, where resources are binned by qubit. The timing
+    between instructions within each circuit schedule are respected.
+
+    Args:
+        circuit: The quantum circuit to translate
+        schedule_config: Backend specific parameters used for building the Schedule
+    Returns:
+        A final schedule, pulses occuring as early as possible
+    """
+    circ_pulse_defs = translate_gates_to_pulse_defs(circuit, schedule_config)
+
+    def update_times(inst_qubits: List[int], time: int = 0) -> None:
+        """Update the time tracker for all inst_qubits to the given time."""
+        for q in inst_qubits:
+            qubit_time_available[q] = time
+
+    sched = Schedule()
+    qubit_time_available = defaultdict(int)
+
+    for circ_pulse_def in circ_pulse_defs:
+        time = max(qubit_time_available[q] for q in circ_pulse_def.qubits)
+        if isinstance(circ_pulse_def.schedule, Barrier):
+            update_times(circ_pulse_def.qubits, time)
+        else:
+            sched |= circ_pulse_def.schedule << time
```

Comment:
I believe that by scheduling in this way we are relying on the `union` operation which is known to be slower and also ignoring channel buffer constraints. Should be an `append` operation instead? What I think makes the most sense is to main a `Schedule` for each barrier and track which qubits belong to which barrier as time proceeds. Finally after processing all instructions `append` all of the barrier schedules together. This offloads the need to track time onto the individual schedules and automatically includes buffering.

Reply:
if we make the upcoming change unions should be as fast as anything else. Buffer times are being ignored, that's a bug here (assuming buffer times should be added here, I still don't really get them). As far as relying on append rather than explicitly tracking time, think about this:

```
qc.x(q[0])
qc.measure(q[0], mem_slots_etc)
```
These instructions don't share any channels, so appending the measure to the X-Schedule would yield a Schedule with both instructions starting at time=0.

Reply:
I just think it is a more natural way to construct circuit from their buffered components. I believe implementing it this way will be very graceful.

Reply:
Sure, I'll try doing it that way